### PR TITLE
Add types for openseadragon

### PR DESCRIPTION
**Pull request by alvaromartmart**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
Please split out the unrelated changes to `cli-progress`

**Comment by alvaromartmart** [comment]
@RyanCavanaugh you're right. I removed that commit. It should be OK now, thanks!

**Comment by kristinruben** [comment]
> Please split out the unrelated changes to `cli-progress`

I think this has been taken care of! Thanks so much.

**Comment by alvaromartmart** [comment]
Hi @RyanCavanaugh, I think I completed the requested change, but I've noticed the PR is still under the "Needs author attention" state in the Status Board... did I miss something?

If this is expected & you have just not had the time yet to review it, feel free to skip this comment and review it whenever you have the time. Thanks!

File: types/openseadragon/index.d.ts
Comment (peterblazejewicz) [inline types/openseadragon/index.d.ts]: Just verified, this could be `3.1`, to not limit TS support to most recent ones

File: types/openseadragon/index.d.ts
Comment (peterblazejewicz) [inline types/openseadragon/index.d.ts]: I think you don't need this line at all, as there is no direct reference to Node I think

File: types/openseadragon/index.d.ts
Comment (peterblazejewicz) [inline types/openseadragon/index.d.ts]: Should this definition for `TileSource` shape contains a `type` property somewhere? Event optional?
There are sampels of code snips that contains type:
https://openseadragon.github.io/examples/tilesource-image/
, the project sample code does not compile due to this:
```ts
const viewer3 = OpenSeadragon({
    id: "openseadragon",
    prefixUrl: "openseadragon/images/",
    showNavigator: false,
    tileSources: {
      type: 'legacy-image-pyramid',
      levels:[{
          url: '2003rosen1799/0001q.jpg',
          height: 889,
          width:  600
      },{
          url: '2003rosen1799/0001r.jpg',
          height: 2201,
          width:  1485
      },{
          url: '2003rosen1799/0001v.jpg',
          height: 4402,
          width:  2970
      }]
    }
  });
```

**Comment by peterblazejewicz** [comment]
@alvaromartmart 
what's the expected type of consumption of this definition type? 
I assume the original is shipped as the UMD module, with the `OpenSeadragon` export as the namespace, so it can be used that way:

```ts
const viewer3 = OpenSeadragon({});
const viewer = new OpenSeadragon.Viewer({ id: 'viewerid' });
const viewport = new OpenSeadragon.Viewport({ margins: {} });
const element = new Element();
const drawer = new OpenSeadragon.Drawer({ viewer, viewport, element });
const mouseTracker = new OpenSeadragon.MouseTracker({ element });
```
or:
```ts
import OpenSeadragon = require('openseadragon');;

const viewer3 = OpenSeadragon({});
const viewer = new OpenSeadragon.Viewer({ id: 'viewerid' });
const viewport = new OpenSeadragon.Viewport({ margins: {} });
```
or event that one (with interop)::
```ts
import OpenSeadragon, { Viewer, Viewport, Drawer, MouseTracker } from 'openseadragon';

const viewer3 = OpenSeadragon({});
const viewer = new Viewer({ id: 'viewerid' });
const viewport = new Viewport({ margins: {} });
const element = new Element();
const drawer = new Drawer({ viewer, viewport, element });
const mouseTracker = new MouseTracker({ element });
```
Thanks!

File: types/openseadragon/index.d.ts
Comment (alvaromartmart) [reply types/openseadragon/index.d.ts]: Changed this, thanks!

File: types/openseadragon/index.d.ts
Comment (alvaromartmart) [reply types/openseadragon/index.d.ts]: :+1:

File: types/openseadragon/index.d.ts
Comment (alvaromartmart) [reply types/openseadragon/index.d.ts]: :+1: you're right, I also added the example to the tests

**Comment by alvaromartmart** [comment]
> @alvaromartmart
> what's the expected type of consumption of this definition type?
> I assume the original is shipped as the UMD module, with the `OpenSeadragon` export as the namespace, so it can be used that way:
> 
> ```ts
> const viewer3 = OpenSeadragon({});
> const viewer = new OpenSeadragon.Viewer({ id: 'viewerid' });
> const viewport = new OpenSeadragon.Viewport({ margins: {} });
> const element = new Element();
> const drawer = new OpenSeadragon.Drawer({ viewer, viewport, element });
> const mouseTracker = new OpenSeadragon.MouseTracker({ element });
> ```
> 
> or:
> 
> ```ts
> import OpenSeadragon = require('openseadragon');;
> 
> const viewer3 = OpenSeadragon({});
> const viewer = new OpenSeadragon.Viewer({ id: 'viewerid' });
> const viewport = new OpenSeadragon.Viewport({ margins: {} });
> ```
> 
> or event that one (with interop)::
> 
> ```ts
> import OpenSeadragon, { Viewer, Viewport, Drawer, MouseTracker } from 'openseadragon';
> 
> const viewer3 = OpenSeadragon({});
> const viewer = new Viewer({ id: 'viewerid' });
> const viewport = new Viewport({ margins: {} });
> const element = new Element();
> const drawer = new Drawer({ viewer, viewport, element });
> const mouseTracker = new MouseTracker({ element });
> ```
> 
> Thanks!

@justincy helped with that. I think it's ok now.

**Comment by peterblazejewicz** [comment]
@alvaromartmart 
It still does not comply with non CommonJS syntax,
can you try something similar to these ones?:
https://gist.github.com/peterblazejewicz/d76a27778e2125cd155dfc679184cf36

**Comment by alvaromartmart** [comment]
> @alvaromartmart
> It still does not comply with non CommonJS syntax,
> can you try something similar to these ones?:
> https://gist.github.com/peterblazejewicz/d76a27778e2125cd155dfc679184cf36

Thank you Peter, that's better. I also added you to the list of collaborators, if that's ok with you.

**Review by peterblazejewicz — APPROVED** [review]
@alvaromartmart 
OK, as this is the first draft, it seems reasonable OK at the moment, for example to work with a code like that one: https://stackblitz.com/edit/typescript-openseadragon?file=index.ts used in the tests. The shape of the definition file can be reworked later to match all usage scenarios :)
LGTM!

**Comment by alvaromartmart** [comment]
Thank you for your help @peterblazejewicz! What would be the next steps, then? The PR still seems blocked by the changes requested by @RyanCavanaugh a while ago, and I don't know how to "re-request" his review again

**Comment by peterblazejewicz** [comment]
@alvaromartmart easy, PRs are reviewed by other MS team folks on regular basis,
keep me in touch

**Comment by alvaromartmart** [comment]
Hi @peterblazejewicz, this is still stuck here. I guess @RyanCavanaugh didn't get notified when I submitted the requested changes, and I don't know how to move forward. Any tips?

**Comment by iangilman** [comment]
I'm the maintainer for the OpenSeadragon library… Just throwing my hat in here in case the reviewers have any questions for me :-)

**Review by andrewbranch — CHANGES_REQUESTED** [review]
Sorry this PR fell under the radar. Thanks @peterblazejewicz for mentioning it in the Gitter.

I see some previous discussion about the module format, but it doesn’t look like it was ever resolved correctly. I still see `export default OpenSeadragon`, which appears to be wrong. @peterblazejewicz’s gist (https://gist.github.com/peterblazejewicz/d76a27778e2125cd155dfc679184cf36) looks correct to me, so I’m not sure why that suggestion wasn’t taken? Am I missing something?

**Comment by alvaromartmart** [comment]
Hi @andrewbranch, thank you for reviewing this. You're right, I actually missed some of @peterblazejewicz suggestions. I just pushed a new commit, I hope I got it right this time!

**Review by andrewbranch — APPROVED** [review]
Thanks for your patience and the updates!

**Comment by alvaromartmart** [comment]
Great!!! Thank you for your help

**Pull request merged**
